pkg/nsx/services/ippool/ippool.go

```diff
@@ -0,0 +1,252 @@
+package ippool
+
+import (
+	"sync"
+	"time"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha2"
+	commonctl "github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var (
+	log                           = logger.Log
+	MarkedForDelete               = true
+	EnforceRevisionCheckParam     = false
+	ResourceTypeIPPool            = common.ResourceTypeIPPool
+	ResourceTypeIPPoolBlockSubnet = common.ResourceTypeIPPoolBlockSubnet
+	NewConverter                  = common.NewConverter
+)
+
+type IPPoolService struct {
+	common.Service
+	ipPoolStore            *IPPoolStore
+	ipPoolBlockSubnetStore *IPPoolBlockSubnetStore
+}
+
+func InitializeIPPool(service common.Service) (*IPPoolService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(2)
+
+	ipPoolService := &IPPoolService{Service: service}
+	ipPoolService.ipPoolStore = &IPPoolStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPPoolCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBindingType(),
+	}}
+	ipPoolService.ipPoolBlockSubnetStore = &IPPoolBlockSubnetStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeIPPoolCRUID: indexFunc}),
+		BindingType: model.IpAddressPoolBlockSubnetBindingType(),
+	}}
+
+	go ipPoolService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPPool, ipPoolService.ipPoolStore)
+	go ipPoolService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPPoolBlockSubnet, ipPoolService.ipPoolBlockSubnetStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return ipPoolService, err
+	}
+	return ipPoolService, nil
+}
+
+func (service *IPPoolService) CreateOrUpdateIPPool(obj *v1alpha2.IPPool) (bool, bool, error) {
+	nsxIPPool, nsxIPSubnets := service.BuildIPPool(obj)
+	if nsxIPPool == nil || len(nsxIPSubnets) == 0 {
```

Comment:
is there possibility that nsxIPPool nil or nsxIPSubnets empty? BTW,  is it supported that https://github.com/vmware-tanzu/nsx-operator/blob/main/pkg/apis/v1alpha2/ippool_types.go#L40 , there is no SubnetRequest in IPPool CRD?

Reply:
Thanks, there's no possibility that nsxIPPool can be nil, however, nsxIPSubnets can be empty since IPPool.Spec.Subnets are optional, I think it's better to remove this check.

Reply:
Done, support when IPPool.Spec.Subnets is nil.